msal/src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -394,107 +406,46 @@ private MsalAccessTokenCacheItem FindAccessTokenCommon
+                    msg = "No tokens found for matching authority, client_id, user and scopes.";
                     requestParams.RequestContext.Logger.Info(msg);
                     requestParams.RequestContext.Logger.InfoPii(msg);
+                    return null;
+                }
 
-                    //authority was passed in the API
-                    IEnumerable<MsalAccessTokenCacheItem> filteredByPrefferedAlias = 
-                        filteredItems.Where
-                        (item => item.Environment.Equals(preferredEnvironmentAlias, StringComparison.OrdinalIgnoreCase))
-                        .ToList();
-
-                    if (filteredByPrefferedAlias.Any())
-                    {
-                        filteredItems = filteredByPrefferedAlias;
-                    }
-                    else
-                    {
-                        filteredItems = filteredItems.Where(
-                            item => environmentAliases.Contains(item.Environment) && 
-                            item.TenantId.Equals(requestParams.Authority.GetTenantId(), StringComparison.OrdinalIgnoreCase)).ToList();
-                    }
-
-                    //no match
-                    if (!filteredItems.Any())
-                    {
-                        msg = "No tokens found for matching authority, client_id, user and scopes.";
-                        requestParams.RequestContext.Logger.Info(msg);
-                        requestParams.RequestContext.Logger.InfoPii(msg);
-                        return null;
-                    }
+                //if only one cached token found
+                if (filteredItems.Count() == 1)
+                {
+                    msalAccessTokenCacheItem = filteredItems.First();
+                }
+                else
+                {
+                    msg = "Multiple tokens found for matching authority, client_id, user and scopes.";
+                    requestParams.RequestContext.Logger.Error(msg);
+                    requestParams.RequestContext.Logger.ErrorPii(msg);
 
-                    //if only one cached token found
-                    if (filteredItems.Count() == 1)
-                    {
-                        msalAccessTokenCacheItem = filteredItems.First();
-                    }
-                    else
-                    {
-                        msg = "Multiple tokens found for matching authority, client_id, user and scopes.";
-                        requestParams.RequestContext.Logger.Error(msg);
-                        requestParams.RequestContext.Logger.ErrorPii(msg);
-                        
-                        throw new MsalClientException(MsalClientException.MultipleTokensMatchedError,
-                            MsalErrorMessage.MultipleTokensMatched);
-                    }
+                    throw new MsalClientException(MsalClientException.MultipleTokensMatchedError,
+                        MsalErrorMessage.MultipleTokensMatched);
```

Comment:
Can we pls. inform the developer on what they should do as part of the error? e.g. that they have to clean the cache?

Reply:
It should never happen, combination of authority clientId user and scope is unique key in the cache, 
if it happened it means that cache is corrupted, due to some bug or error...  current  message MultipleTokensMatched = "The cache contains multiple tokens satisfying the requirements. Call AcquireToken again providing more requirements like authority"  is not relevant any more since authotiy is mandatory now, i can change it to "The cache contains multiple tokens satisfying the requirements. Try to clear token cache and report the issue"

Reply:
(not a fan of report the issue part, kindly remove that) if anything point to a page in the wiki where we can describe to the developer how to handle the situation the best.

---
In reply to: [216103209](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1246#discussion_r216103209) [](ancestors = 216103209)